### PR TITLE
add e2e dashboard backend

### PR DESCRIPTION
**Pull request by Aman-Codes**
## Proposed changes

@uditgaurav @Jonsy13  Added the e2e dashboard backend code which was earlier present at https://github.com/Aman-Codes/e2e-dashboard-backend

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [x] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus-e2e/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the commit for DCO to be passed.
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added Screenshots or Terminal snippets in the PR comment to validate the successful execution of changes

**Documentation**
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] data-cy=* parameters have been added in codebase ( ChaosCenter Tests )
- [ ] data-cy=* parameters need to be added in codebase ( ChaosCenter Tests )
- [ ] No updates required.

**Code Review**
- [x] Does the test handle fatal exceptions, ie. uncaught exceptions from 3rd party library (if any)

**Issue**
- [x] Tasks in issue are checked off

## Dependency
- Please add the links to the dependent PR need to be merged before this (if any).

### CODE REVIEW

**Comment by Aman-Codes** [comment]
For testing deployed the code to https://e2edashboardbackend.herokuapp.com/status

File: dashboard-backend/main.go [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/router"
```
Comment (uditgaurav) [inline dashboard-backend/main.go:4]: we can import using litmuschaos
```suggestion
	"github.com/litmuschaos/e2e-dashboard-backend/pkg/router"
```

File: dashboard-backend/pkg/env/env.go [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package env
+
+import (
+	"os"
+
+	"github.com/joho/godotenv"
+	"github.com/litmuschaos/litmus-go/pkg/log"
+)
+
+func GoDotEnvVariable(key string) string {
```
Comment (uditgaurav) [inline dashboard-backend/pkg/env/env.go:10]: can we put comments on every func like:

```
// GoDotEnvVariable <your-comment>
func GoDotEnvVariable(key string) string {
```

File: dashboard-backend/pkg/env/env.go [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+	// load .env file
+	err := godotenv.Load(".env")
+	if err != nil {
+		log.Fatalf("Error loading .env file")
```
Comment (uditgaurav) [inline dashboard-backend/pkg/env/env.go:14]: 1. Do we need that comment?
2. log.Fatalf will do exit 1 and the further execution will stop so can we use `log.Errorf` instead?

File: dashboard-backend/pkg/fetchLog/fetchLog.go [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/constants"
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/customErrors"
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/env"
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/unzip"
+	"github.com/Aman-Codes/e2e-dashboard-backend/pkg/utils"
```
Comment (uditgaurav) [inline dashboard-backend/pkg/fetchLog/fetchLog.go:15]: same as above for all import let's use litmuschaos

File: dashboard-backend/pkg/fetchLog/fetchLog.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	JobName    string `json:"jobName" binding:"required"`
+	StepNumber string `json:"stepNumber" binding:"required"`
+}
+
+func parseUrl(fullURLFile string) (string, error) {
+	log.Info("Start to fetch log")
+	fileURL, err := url.Parse(fullURLFile)
+	if err != nil {
+		log.Errorf("failed to parse URL, err %v", err)
+		return "", customErrors.InternalServerError()
```
Comment (uditgaurav) [inline dashboard-backend/pkg/fetchLog/fetchLog.go:31]:
```suggestion
		return "", errors.Errorf( "failed to parse URL, err %v", customErrors.InternalServerError())
```
Please check this for other places as well

**Review by uditgaurav — CHANGES_REQUESTED** [review]

File: dashboard-backend/pkg/fetchLog/fetchLog.go (hunk 5 above)
Comment (Aman-Codes) [reply dashboard-backend/pkg/fetchLog/fetchLog.go:32]: Could you please tell which `errors` package are you referring to?
As using `log.Errorf` gives the below error
image

File: dashboard-backend/go.mod [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	github.com/sirupsen/logrus v1.7.0 // indirect
+	github.com/ugorji/go/codec v1.2.6 // indirect
+	golang.org/x/crypto v0.0.0-20210921155107-089bfa567519 // indirect
+	golang.org/x/sys v0.0.0-20211031064116-611d5d643895 // indirect
+	golang.org/x/text v0.3.7 // indirect
+	google.golang.org/protobuf v1.27.1 // indirect
+	gopkg.in/yaml.v2 v2.4.0 // indirect
+)
+
+replace (
```
Comment (Jonsy13) [inline dashboard-backend/go.mod:33]: @Aman-Codes We don't require these, please use `go mod tidy` before committing the changes. cc: @uditgaurav

Comment (uditgaurav) [reply dashboard-backend/go.mod:33]: you can minimise and run `go mod tidy`

**Review by uditgaurav — APPROVED** [review]

File: dashboard-backend/pkg/fetchLog/fetchLog.go (hunk 5 above)
Comment (uditgaurav) [reply dashboard-backend/pkg/fetchLog/fetchLog.go:33]: okay will update it

File: dashboard-backend/go.mod (hunk 6 above)
Comment (Aman-Codes) [reply dashboard-backend/go.mod:33]: On removing any `replace` I get the following errors
image
cc @Jonsy13 @uditgaurav

Comment (Aman-Codes) [reply dashboard-backend/go.mod:33]: running `go mod tidy` didn't have any change

**Review by Jonsy13 — APPROVED** [review]

File: dashboard-backend/pkg/fetchLog/fetchLog.go [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/litmuschaos/e2e-dashboard-backend/pkg/constants"
+	"github.com/litmuschaos/e2e-dashboard-backend/pkg/customErrors"
+	"github.com/litmuschaos/e2e-dashboard-backend/pkg/env"
+	"github.com/litmuschaos/e2e-dashboard-backend/pkg/unzip"
+	"github.com/litmuschaos/e2e-dashboard-backend/pkg/utils"
```
Comment (uditgaurav) [inline dashboard-backend/pkg/fetchLog/fetchLog.go:17]: The imports should be `github.com/litmuschaos/litmuse2e/e2e-dashboard-backend/pkg/constants`

**Review by uditgaurav — CHANGES_REQUESTED** [review]

File: dashboard-backend/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+Create a file `.env` and add the below credentials in that file.
```
Comment (uditgaurav) [inline dashboard-backend/README.md:1]: Can we provide a little more description of this, describing what, why, and how?

Comment (Aman-Codes) [reply dashboard-backend/README.md]: @uditgaurav Added detailed installation steps

File: dashboard-backend/pkg/fetchLog/fetchLog.go (hunk 7 above)
Comment (Aman-Codes) [reply dashboard-backend/pkg/fetchLog/fetchLog.go]: @uditgaurav Updated the imports

**Pull request merged**